### PR TITLE
Related Entity Location Sync Strategy.

### PR DESCRIPTION
**Pull request by lincmba**
- User selected sync strategy
 - User selected sync strategy fixes https://github.com/opensrp/fhircore/issues/3052

**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Resolves [link to issue]

**Engineer Checklist**

- [x] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [x] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [x] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [x] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

**Comment by pld** [comment]
nice! @ndegwamartin can you please prioritize review of this so we can get it in asap, thank you!

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java [hunk 1]
```diff
@@ -20,19 +20,19 @@ public class Constants {
...
-
     public static final int PAGINATION_DEFAULT_PAGE_SIZE = 20;
-
     public static final int PAGINATION_DEFAULT_PAGE_NUMBER = 1;
+    public static final String SYNC_LOCATIONS = "_syncLocations";
+    public static final String RELATED_ENTITY_LOCATION = "RelatedEntityLocation";
+    public static final String RELATED_ENTITY_TAG_URL_ENV = "RELATED_ENTITY_TAG_URL";
+    public static final String DEFAULT_RELATED_ENTITY_TAG_URL =
+            "https://smartregister.org/related-entity-location-tag-id";
+    public static final String ROLE_ALL_LOCATIONS = "ROLE_ALL_LOCATIONS";
```
Comment (lincmba) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java:35]: This is the name of the role as described in the issue. I don't think it is standard for keycloak roles to be preceded by `ROLE`. Should it he kept as `ROLE_ALL_LOCATIONS` or `ALL_LOCATIONS`?

Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Constants.java:35]: Good question. I'm okay with dropping it or if we need a prefix I'd go with `OPENSRP_ALL_LOCATIONS`. But I'm thinking if we do that then we need to prefix all our roles for consistency. Dropping it might be best for now cc @pld @dubdabasoduba

File: plugins/src/test/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecisionTest.java [hunk 2]
```diff
@@ -92,6 +95,127 @@ public void preProcessShouldAddLocationIdFiltersWhenUserIsAssignedToLocationsOnl
         }
     }
 
+    @Test
+    public void
+            preProcessShouldAddAssignedRelatedEntityLocationIdsFiltersWhenUserIsAssignedToRelatedEntityLocationIdsOnly()
```
Comment (ndegwamartin) [inline plugins/src/test/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecisionTest.java:100]: We can rename the test case to conform to the format `<methodName><ConditionUnderTest><ExpectedBehavior>`

See https://docs.opensrp.io/engineering/android-app/developer-setup/testing/unit-tests

File: plugins/src/test/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecisionTest.java [hunk 3]
```diff
@@ -92,6 +95,127 @@ public void preProcessShouldAddLocationIdFiltersWhenUserIsAssignedToLocationsOnl
...
+
+        for (String param : mutatedRequest.getQueryParams().get(Constants.TAG_SEARCH_PARAM)) {
+            Assert.assertFalse(param.contains(Constants.DEFAULT_CARE_TEAM_TAG_URL));
+            Assert.assertFalse(param.contains(Constants.DEFAULT_ORGANISATION_TAG_URL));
+        }
+    }
+
+    @Test
+    public void
+            preProcessShouldAddSelectedRelatedEntityLocationIdsFiltersWhenUserIsAssignedToRelatedEntityLocationIdsOnly()
```
Comment (ndegwamartin) [inline plugins/src/test/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecisionTest.java:141]: Same as above

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java [hunk 4]
```diff
@@ -546,7 +546,7 @@ private List<OrganizationAffiliation> mapBundleToOrganizationAffiliation(
                 : Collections.emptyList();
     }
 
-    private List<LocationHierarchy> getLocationsHierarchy(List<String> locationsIdentifiers) {
+    static List<LocationHierarchy> getLocationsHierarchy(List<String> locationsIdentifiers) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java:549]: You can expand the accessibility by updating the access modifier to `public`
```suggestion
    public static List<LocationHierarchy> getLocationsHierarchy(List<String> locationsIdentifiers) {
```

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java [hunk 5]
```diff
@@ -43,7 +43,7 @@ public class PractitionerDetailsEndpointHelper {
     public static final Bundle EMPTY_BUNDLE = new Bundle();
     private final IGenericClient r4FHIRClient;
 
-    private final LocationHierarchyEndpointHelper locationHierarchyEndpointHelper;
+    private static LocationHierarchyEndpointHelper locationHierarchyEndpointHelper;
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java:46]: This field doesn't seem to be used anywhere else in this class. Let's remove this field and instantiate it at the only point of usage.

**Note:** For this to work you will need to refactor `IGenericClient getFhirClientForR4() `method to static so that it can be used in the constructor of `LocationHierarchyEndpointHelper`

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java [hunk 6]
```diff
@@ -413,6 +414,18 @@ private static Map<String, List<String>> collateSyncStrategyIds(
                                     : new ArrayList<>();
 
                     resultMap = Map.of(syncStrategy, locationIds);
+                } else if (Constants.RELATED_ENTITY_LOCATION.equalsIgnoreCase(syncStrategy)) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java:417]: This should be a separate if statement _(rather than else if)_ so that incase the newly processed `this.syncStrategyIds` is empty the exception is still thrown and logged.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 7]
```diff
@@ -101,8 +101,25 @@ public RequestMutation getRequestMutation(RequestDetailsReader requestDetailsRea
...
-            if (syncStrategyIds.isEmpty()
+            if (Constants.RELATED_ENTITY_LOCATION.equalsIgnoreCase(syncStrategy)) {
+                Map<String, String[]> parameters =
+                        new HashMap<>(requestDetailsReader.getParameters());
+                String[] syncLocations = parameters.get(Constants.SYNC_LOCATIONS);
+                List<String> relatedEntityLocationIds;
+                if (roles.contains(Constants.ROLE_ALL_LOCATIONS) && syncLocations != null) {
+                    // selected locations
+
+                    List<String> locationUuids = Arrays.asList(syncLocations[0].split(","));
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:114]: You can use the existing `Constants.PARAM_VALUES_SEPARATOR` instead of comma

Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:114]: For better performance we can can use a `Set` to remove any duplicates.

```suggestion
                    Set<String> locationUuids = Set.of(syncLocations[0].split(","));
```

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 8]
```diff
@@ -101,8 +101,25 @@ public RequestMutation getRequestMutation(RequestDetailsReader requestDetailsRea
...
                         requestDetailsReader)) { // Check if it is the Sync URL and Skip app-wide
+
             // accessible resource requests
-            if (syncStrategyIds.isEmpty()
+            if (Constants.RELATED_ENTITY_LOCATION.equalsIgnoreCase(syncStrategy)) {
+                Map<String, String[]> parameters =
+                        new HashMap<>(requestDetailsReader.getParameters());
+                String[] syncLocations = parameters.get(Constants.SYNC_LOCATIONS);
+                List<String> relatedEntityLocationIds;
+                if (roles.contains(Constants.ROLE_ALL_LOCATIONS) && syncLocations != null) {
```
[the remark this answers is not in the document]
Comment (pld) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:111]: good call, this definitely needs a test, I'd write a failing test first, then fix the code here

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java (hunk 5 above)
Comment (lincmba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java:46]: It is used here https://github.com/onaio/fhir-gateway-extension/blob/related_entity_location_sync_strategy/plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PractitionerDetailsEndpointHelper.java#L50

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 9]
```diff
@@ -101,8 +101,25 @@ public RequestMutation getRequestMutation(RequestDetailsReader requestDetailsRea
...
+
+                    List<String> locationUuids = Arrays.asList(syncLocations[0].split(","));
+                    relatedEntityLocationIds =
+                            PractitionerDetailsEndpointHelper.getAttributedLocations(
+                                    PractitionerDetailsEndpointHelper.getLocationsHierarchy(
+                                            locationUuids));
+                    this.syncStrategyIds = Map.of(syncStrategy, relatedEntityLocationIds);
+                }
+
+            } else if (syncStrategyIds.isEmpty()
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:122]: This should be a separate if statement _(rather than else if)_ so that incase the newly processed `this.syncStrategyIds` is empty the exception is still thrown and logged.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 10]
```diff
@@ -4,11 +4,7 @@
 
 import java.io.FileReader;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:7]: Revert from using wildcard importation for this

**Review by ndegwamartin — APPROVED** [review]
